src/librustdoc/html/render/mod.rs

```diff
@@ -1112,28 +1115,75 @@ fn render_assoc_items<'a, 'cx: 'a>(
     containing_item: &'a clean::Item,
     it: DefId,
     what: AssocItemRender<'a>,
+    aliased_type: Option<DefId>,
 ) -> impl fmt::Display + 'a + Captures<'cx> {
     let mut derefs = DefIdSet::default();
     derefs.insert(it);
     display_fn(move |f| {
-        render_assoc_items_inner(f, cx, containing_item, it, what, &mut derefs);
+        render_assoc_items_inner(f, cx, containing_item, it, what, &mut derefs, aliased_type);
         Ok(())
     })
 }
 
+fn is_valid_impl_for(tcx: TyCtxt<'_>, item_def_id: DefId, impl_def_id: DefId) -> bool {
+    let infcx = tcx.infer_ctxt().intercrate(true).build();
+    let ocx = ObligationCtxt::new(&infcx);
+    let param_env = ParamEnv::empty();
+
+    let alias_substs = infcx.fresh_substs_for_item(rustc_span::DUMMY_SP, item_def_id);
+    let alias_ty = tcx.type_of(item_def_id).subst(tcx, alias_substs);
+    let alias_bounds = tcx.predicates_of(item_def_id).instantiate(tcx, alias_substs);
+
+    let impl_substs = infcx.fresh_substs_for_item(rustc_span::DUMMY_SP, impl_def_id);
+    let impl_self_ty = tcx.type_of(impl_def_id).subst(tcx, impl_substs);
+    let impl_bounds = tcx.predicates_of(impl_def_id).instantiate(tcx, impl_substs);
+
+    if ocx.eq(&ObligationCause::dummy(), param_env, impl_self_ty, alias_ty).is_err() {
+        return false;
+    }
+    ocx.register_obligations(
+        alias_bounds
+            .iter()
+            .chain(impl_bounds)
+            .map(|(p, _)| Obligation::new(tcx, ObligationCause::dummy(), param_env, p)),
+    );
+
+    let errors = ocx.select_where_possible();
+    errors.is_empty()
+}
+
+// If `aliased_type` is `Some`, it means `it` is a type alias and `aliased_type` is the "actual"
+// type aliased behind `it`. It is used to check whether or not the implementation of the aliased
+// type can be displayed on the alias doc page.
 fn render_assoc_items_inner(
     mut w: &mut dyn fmt::Write,
     cx: &mut Context<'_>,
     containing_item: &clean::Item,
     it: DefId,
     what: AssocItemRender<'_>,
     derefs: &mut DefIdSet,
+    aliased_type: Option<DefId>,
```

Comment:
We need a variable to tell us that this is an alias. From this, since we already have the `DefId`, why not using it directly?

Reply:
I edited my comment, I don't how but I misread your comment the first time. Sorry about that.

Reply:
Ah found it! I need it because of `Some(aliased_type) => cache.impls.get(&aliased_type).unwrap_or(&empty),`. We can't retrieve the `DefId` easily from a `Ty`. The only method I found to get it is [ty_adt_id](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/query/keys/trait.Key.html#method.ty_adt_id) but it doesn't cover all the cases we need.

Reply:
Having `aliased_type` separately here feels off to me and may at least deserve a comment, but that's unrelated to the type system itself so I am fine with that.

Reply:
The doc comment for `aliased_type` is already on the function. ;)